### DISCUSSION TITLE
Config not being picked up from after/ftplugin/rust.lua

### DISCUSSION THREAD

**Discussion by adaszko** [mrcjkb/rustaceanvim#77] [Q&A] [answered]
### Neovim version (nvim -v)

v0.9.4

### Operating system/version

macOS 14.1.1

### Output of `:checkhealth rustaceanvim`

============================================================================== rustaceanvim: require("rustaceanvim.health").check()  Checking for Lua dependencies ~ - OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.  Checking external dependencies ~ - OK rust-analyzer: found rust-analyzer 0.3.1713-standalone (f49320774 2023-10-28) - OK Cargo: found cargo 1.73.0 (9c4383fb5 2023-08-26) - OK rustc: found rustc 1.73.0 (cc66ad468 2023-10-03) - OK debug adapter: found codelldb   Checking config ~ - OK vim.g.rustaceanvim is set - OK No errors found in config.  Checking for conflicting plugins ~ - OK No conflicting plugins detected.

### How to reproduce the issue

Open any Rust source file.  Type `foo.`.  Observe the list of completions which is missing some items.

### Expected behaviour

Typing `foo.rcc<TAB>` should expand to `Rc::clone(&foo)`.

### Actual behaviour

The `rcc` postfix snippet completion is completely absent from nvim-cmp completion list.  The built-in rust-analyzer postfix completions like e.g. `foo.arc` are present though and this configuration worked with rust-tools.

### The minimal config used to reproduce this issue.

In my `~/.config/nvim/after/ftplugin/rust.lua`:

```
vim.g.rustaceanvim = {
  server = {
    settings = {
      ["rust-analyzer"] = {
        -- https://github.com/rust-lang/rust-analyzer/blob/master/docs/user/generated_config.adoc
        completion = {
          snippets = {
            custom = {
                ["Rc::clone"] = {
                    postfix = "rcc",
                    body = "Rc::clone(&${receiver})",
                    requires = "std::rc::Rc",
                    scope = "expr"
                },
            },
          },
        },
      },
    },
  },
}
```

**Comment by mrcjkb** [comment]
Hey 👋 

You don't want to set `vim.g.rustaceanvim` in `after/ftplugin/rust.lua`. This causes it to be sourced after starting the LSP client, so it won't be picked up.

**Reply by adaszko** [reply]
Hi.  Thanks for taking a look.  While I was able to confirm that placing snippets in rust-analyzer.json file does work, the exact same single snippet config copied over to .vimrc settings section does not work.

**Reply by mrcjkb** [reply]
I just gave it a try and it works for me:

![scrot](https://github.com/mrcjkb/rustaceanvim/assets/12857160/edb52318-6254-4d02-9ed9-03356c9fe100)

Do you have a link to your config?

Btw, if you run `:checkhealth rustaceanvim` after updating to the latest version of the plugin, it will output an error if you have `vim.g.rustaceanvim` set, but it wasn't picked up during initialization.

**Reply by adaszko** [reply]
Yup, I'm getting that error: `  - ERROR vim.g.rustaceanvim is set, but it was sourced after rustaceanvim was initialized.`

**Reply by adaszko** [reply]
My config isn't public but I'm doing this, in ~/.vimrc:

```
if has("nvim")
lua <<EOF
vim.g.rustaceanvim = {
 ...
}
EOF
endif
```

**Reply by mrcjkb** [reply]
🤔 That's very strange.

Are you calling `require('rustaceanvim')` or any other rustaceanvim Lua modules anywhere in your config?

Something seems to be causing the plugin to be initialised before your config is sourced. Normally, it should only initialise when you open a rust file.

**Reply by adaszko** [reply]
Nope, I'm not calling any of that

**Reply by mrcjkb** [reply]
There's not much I can do without being able to look at your config.
I would suggest adding `print(debug.traceback())` at the beginning of [this module](https://github.com/mrcjkb/rustaceanvim/blob/master/lua/rustaceanvim/config/internal.lua) in your local checkout.
Then when you start neovim, the stack trace should tell you what is causing the module to be loaded.
Feel free to post it here - maybe I can help if it's too cryptic.

**Reply by mrcjkb** [reply]
I just remembered, there was a [similar issue](https://github.com/mrcjkb/haskell-tools.nvim/pull/226#issuecomment-1673131969) someone had with haskell-tools (a similar filetype plugin, but for Haskell). Something in their `.vimrc` was causing the plugin to be initialised eagerly (perhaps a neovim bug?).

We never identified the culprit, but maybe printing `debug.traceback` will help this time.

**Reply by adaszko** [reply]
Here's what I got:

```
stack traceback:
^I.../start/rustaceanvim/lua/rustaceanvim/config/internal.lua:1: in main chunk
^I[C]: in function 'require'
^I...ig/nvim/pack/minpac/start/rustaceanvim/ftplugin/rust.lua:2: in main chunk
^I[C]: in function 'nvim_cmd'
^I...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:25: in function <...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:24>
^I[C]: in function 'nvim_buf_call'
```

There's just one call to it.  I'm using neovim on macOS and minpac as the package manager.

**Reply by adaszko** [reply]
My .vimrc has ~2300 lines so looking for a bug in it is like searching a needle in a haystack.  I tried moving the assignment to `vim.g.rustaceanvim = ...` to the very top of it and it didn't change a thing, so I think nothing is overwriting it.

**Reply by adaszko** [reply]
The traceback above comes from opening a rust file via `nvim file.rs`.  However, when I open a file in a running nvim, it's different:

```
stack traceback:
^I.../start/rustaceanvim/lua/rustaceanvim/config/internal.lua:1: in main chunk
^I[C]: in function 'require'
^I...ig/nvim/pack/minpac/start/rustaceanvim/ftplugin/rust.lua:2: in main chunk
^I[C]: in function 'nvim_cmd'
^I...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:25: in function <...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:24>
^I[C]: in function 'nvim_buf_call'
^I...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:24: in function <...brew/Cellar/neovim/0.9.4/share/nvim/runtime/filetype.lua:10>
stack traceback:
^I[string ":lua"]:28: in function <[string ":lua"]:1>
^I[C]: in function 'pcall'
^I...w/Cellar/neovim/0.9.4/share/nvim/runtime/lua/vim/lsp.lua:1538: in function '_on_attach'
^I...w/Cellar/neovim/0.9.4/share/nvim/runtime/lua/vim/lsp.lua:1357: in function ''
^Ivim/_editor.lua: in function <vim/_editor.lua:0>
```

And this time, my `on_attach` code is in effect (keybindings defined there are working) and `healthcheck` says `OK`!

I thought it may be a race condition based on these observations so I replaced the `print(traceback...` with `+ os.execute("sleep " .. tonumber(5))` but it went back to being broken.

**Reply by adaszko** [reply]
To summarise for clarity: the plugin doesn't work when opening a Rust file like `nvim file.rs`.  Opening the file in a running nvim works fine.

**Reply by mrcjkb** [reply]
Interesting :thinking: . I was just able to reproduce this.

It looks like this may be a Neovim bug:
Your `start` plugins are loaded and filetype detection is enabled before your `.vimrc` is sourced - [which shouldn't be happening - see steps 7 & 8 of the neovim initialization startup](https://neovim.io/doc/user/starting.html#initialization).

I will investigate this further when I have some more time, and then file a bug report.

In the meantime, you can work around this (with minpac) as follows:

- Set the plugin type to `opt`:

```vim
call minpac#add('mrcjkb/rustaceanvim', {'type': 'opt'})
```

- Call `packadd! rustaceanvim` after setting `vim.g.rustaceanvim`

```vim
if has("nvim")
lua <<EOF
vim.g.rustaceanvim = {
 ...
}
EOF
packadd! rustaceanvim
endif
```

**Reply by mrcjkb** [reply]
After some initial investigation:

- I can reproduce the issue with `start` plugins using Nix (which calls `set packpath` and `set rtp` before sourcing the user config - and I presume minpac does, too).
- I cannot reproduce the issue by cloning the plugin into Neovim's default `packpath/<somedir>/start` directory.

**Reply by mrcjkb** [reply]
I've created https://github.com/k-takata/minpac/issues/152, and will do some more investigation with the Nix derivation.

**Reply by adaszko** [reply]
Amazing!  Can confirm the workaround does work! 🎉🎉🎉

**Comment by mrcjkb** [comment, marked as answer]
I've been able to reproduce this behaviour for `start` plugins, and traced it down to a call to `syntax on` or `syntax enable`, causing start ftplugins on the packpath to be sourced prematurely.

See https://github.com/mrcjkb/neovim-syntax-bug

I couldn't reproduce it with vim, so I believe it's a possible neovim bug.

**Reply by adaszko** [reply]
Huh, I did in fact found this too via halving of vimrc but I didn't have proof it's a neovim bug. Interesting.  Thanks again for taking a look.

**Reply by mrcjkb** [reply]
I don't have 100 % proof that it's a neovim bug yet, but I'll pick up work on it in a few days 😀

**Reply by mrcjkb** [reply]
https://github.com/neovim/neovim/issues/26859

**Reply by mrcjkb** [reply]
Related: https://github.com/neovim/neovim/issues/19008
